### DISCUSSION TITLE
CoreDNS + external plugins build automation

### DISCUSSION THREAD

**Discussion by chrisohaver** [coredns/coredns#4679] [Ideas]
Entertaining the idea of _automating_ builds of CoreDNS with external plugins:

The build process with external plugins is already _partly_ automated via `plugin.cfg`, however it requires manually selecting a position in the list for the external plugin. This isn't always a straightforward decision for users and not all external plugins provide guidance.  Furthermore, it can't be automated/scripted, for example in CI tests.

The process could be fully automated if ...

* External plugins were to define their required/recommended position in the plugin.cfg in, for example, a yaml file in the external plugin's repository. This places the responsibility in the hands of external plugin developers to select an appropriate position in the plugin chain for their plugin.
* A CoreDNS builder (either in coredns/coredns or some external repo), accepts a list of external plugins to add, and adds the listed plugins to CoreDNS based on the yaml file defined in each external plugin repo.

I've put together a [simplistic POC](https://github.com/chrisohaver/coredns/compare/master...chrisohaver:gen-ext-plugs) that does this in the code generation step of CoreDNS (`go generate`). In effect, ...
* Each supported external repo defines a file `plugin.cfg.yaml`, which defines _where_  the plugin should be inserted in relationship to a specified internal plugin . e.g. before cache
*  When generating `zDirectives.go` and `zplugins.go`, it looks for an environment variable that contains a space delimited list of `external/plugin/repos@version` , and inserts them per their respective `plugin.cfg.yaml` definition.

Is the feature something we can support in coredns/coredns? Or is it something that should it live in a separate repository?
Other thoughts?

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "[coredns/coredns] CoreDNS + externa..." ]
>Entertaining the idea of automating builds of CoreDNS with external plugins:
>
>The build process with external plugins is already partly automated via
>plugin.cfg, however it requires manually selecting a position in the list for
>the external plugin. This isn't always a straightforward decision for users and
>not all external plugins provide guidance. Furthermore, it can't be automated/
>scripted, for example in CI tests.
>
>The process could be fully automated if ...
>
>  • External plugins were to define their required/recommended position in the
>    plugin.cfg in, for example, a yaml file in the external plugin's
>    repository. This places the responsibility in the hands of external plugin
>    developers to select an appropriate position in the plugin chain for their
>    plugin.
>  • A CoreDNS builder (either in coredns/coredns or some external repo),
>    accepts a list of external plugins to add, and adds the listed plugins to
>    CoreDNS based on the yaml file defined in each external plugin repo.
>
>I've put together a simplistic POC that does this in the code generation step
>of CoreDNS (go generate). In effect, ...
>
>  • Each supported external repo defines a file plugin.cfg.yaml, which defines
>    where the plugin should be inserted in relationship to a specified internal
>    plugin . e.g. before cache
>  • When generating zDirectives.go and zplugins.go, it looks for an environment
>    variable that contains a space delimited list of external/plugin/
>    ***@***.*** , and inserts them per their respective plugin.cfg.yaml
>    definition.
>
>Is the feature something we can support in coredns/coredns? Or is it something
>that should it live in a separate repository?
>Other thoughts?

I think it will be hard to convince external plugin authors to add a file that says this.
It also needs to be kept up to date with plugin changes from the coredns side - I don't
see that happening.

There is some convience of providing a builder and making it somehow blessed by us, but I
wouldn't trust any 3rd party builder to create binaries for me (yes this includes the
insanity of running a docker container we provide).

A very easy (and maybe lame) idea is to just a couple of comments in plugin.cfg that tell
you 'put your external plugin somewhere here'?

/Miek

--
Miek Gieben

**Comment by chrisohaver** [comment]
Thanks for the feedback.

> I think it will be hard to convince external plugin authors to add a file that says this.

Many external plugins are not actively maintained. But initially just the ones in the `coredns/` project can define this.  For other plugins, we could leave it to open-source community pressure.

> It also needs to be kept up to date with plugin changes from the coredns side - I don't see that happening.

I don't think this would be common for it to need to change, unless a plugin used as a landmark is removed or changes name (plugin removal/rename has happened, but is rare). For new plugins, it's possible for this to actually matter. E.g. if we were to add an alternate implementation of cache called _cache2_ to the plugin chain after _cache_, external plugins ordered after _cache_ probably would need to be updated to be ordered after _cache2_.  Although possible, I think it would be uncommon.  It could prompt us to reserve new plugins for minor releases - which probably makes sense anyways, IMO releasing new plugins in patch releases feels awkward.

This touches on another issue, that the relative ordering of two external plugins may matter, and depending on their relative position to built in plugins, that might not be easily controlled in a static way (e.g. if two external plugins both define they should be after _cache_, then their relative position to each other is dependent on order they were passed to the build process).

>There is some convience of providing a builder and making it somehow blessed by us, but I
wouldn't trust any 3rd party builder to create binaries for me (yes this includes the
insanity of running a docker container we provide).

Yeah, this partly why I lean toward integrating it into the coredns/coredns build process.

> A very easy (and maybe lame) idea is to just a couple of comments in plugin.cfg that tell
you 'put your external plugin somewhere here'?

For external plugins that I maintain, I have been adding a section to the README to explain where the plugin should go in `plugin.cfg`.
If this explanation is defined as relative to other plugins (e.g. "put it after _cache_"), then it is subject to the same uncommon maintenance required by a yaml file.

This does not work for an automated processes such as integration testing.  But I suppose this can be reasonably left to the integration tests to control this on an individual basis when they build coredns with the plugin/plugins they are testing.
